### PR TITLE
ENH Contingency tables

### PR DESCRIPTION
**Pull request by kshedden**
I added several methods for contingency tables.  Some of these are based on code taken from the sandbox while others are all new.

These procedures are often known by the name of their inventor, but I prefer to give names based on what they do.  Stata takes a similar approach.

Here are the new procedures:

symmetry : based on code taken from sandbox ('bowker test'), assesses symmetry of the underlying joint distribution for a square table.

ordinal_association : all new code, assesses association between rows and columns that have an ordinal interpretation.  It implements Agresti's "linear by linear association test", which is a generalized version of the Cochran-Armitage trend test.  Basically a Pearson chi^2 test for ordinal tables.

stratified_association : all new code, implements the "Cochran-Mantel-Haenszel test" and associated estimate and CI's for the common odds ratio in a set of stratified 2x2 tables.

homogeneity : all new code,  compares row and column marginal distributions for a square table.  Generalizes McNemar's test to two factors with more than two levels.

mcnemar : pulled from sandbox, special case of homogeneity test for 2x2 tables

cochrans_q : pulled from sandbox, generalization of McNemar's test to multiple factors with two levels each.

Design issues:

All functions take a contingency table as the input.  There is no facility for passing in the data itself.  The table can be an ndarray or Pandas. 

Most functions have the option of returning multiple return values or a bunch.  In many cases the bunch contains additional information not available otherwise.  There are no proper results classes here.

Test coverage is decent but it uses "nice" data sets.  There is little testing for corner cases or pathological examples (e.g. all zeros, non-integers, negative values).

Missing pieces:

This is a large area so this is far from complete.  There is very little "exact inference" here, and no special facilities for large and/or sparse tables.

References:

R "coin" package : http://cran.r-project.org/web/packages/coin/coin.pdf

Stata : http://www.stata.com/manuals13/rsymmetry.pdf

SAS : http://www2.sas.com/proceedings/forum2008/382-2008.pdf

R Mantel-Haenszel : https://stat.ethz.ch/R-manual/R-devel/library/stats/html/mantelhaen.test.html

Background about common odds ratios : http://www.biostat.umn.edu/~dipankar/bmtry711.11/lecture_15.pdf
http://www.itl.nist.gov/div898/software/dataplot/refman1/auxillar/mantel.htm

### CODE REVIEW

**Comment by josef-pkt** [comment]
It might be best to import true division for this: `from __future__ import division` to avoid worrying about integer division with counts

@kshedden If you have an overview, then the overview documentation page would be useful, that can also point to the scipy.stats functions fisher exact and independence tests
something like http://statsmodels.sourceforge.net/devel/diagnostic.html#diagnostics

**Comment by kshedden** [comment]
This is now partially refactored into classes.  A class encapsulates several methods that can be applied to the same type of table.  For example, there is a set of procedures that can be applied to a collection of 2x2 tables, so these have now become methods of the ``StratifiedTables`` class.  The other class I currently have is called ``TableSymmetry`` and encapsulates two methods that make sense for square tables with the same categories in the rows and columns. 

Each class has individual methods for calculating the results separately, and a summary method that calculates all results and returns a summary table object.

These aren't model classes so there is no "fit" method, but otherwise they behave like model classes.

**Comment by josef-pkt** [comment]
I briefly browsed the classs, I was worried that it becomes too "heavy" for simple hypothesis tests.

I think it looks good, it's more like Results classes than model classes.
One possibility, since I'm not very familiar with how to do these things in pandas, would be to add a `from_data` `classmethod` that constructs the table for the class.

**Comment by kshedden** [comment]
Agreed, it is intended to be lightweight.  Nothing should be computed until needed.  Since there are quite a few methods for contingency tables, I think it will be helpful to group them according to the high-level analysis goal or the type of input table.  Some of the methods cache intermediate results that can be re-used by other methods acting on the same data.  There are also some "diagnostic" methods to probe, say, where a given test result is coming from.  It is easier to do this with a class than with a bunch of disconnected methods.

The downside is if, say, someone is doing a simulation study and just wants 10,000 p-values, there may be a slight overhead for class creation compared to a simple function call.  But I have tried to make it so that overhead is negligible.

I added some ``from_data`` class methods.  The ``__init__`` interface always takes the contingency table, but ``from_data`` calculates the table for you from data.

File: statsmodels/stats/contingency_tables.py [hunk 1]
```diff
@@ -0,0 +1,1215 @@
...
+
+        self.table_orig = table
+        self.table = np.asarray(table, dtype=np.float64)
+
+        if shift_zeros and (self.table.min() == 0):
+            self.table = self.table + 0.5
+
+
+    @classmethod
+    def from_data(cls, data):
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: this should also have the `shift_zeros=True` option

File: statsmodels/stats/contingency_tables.py [hunk 2]
```diff
@@ -0,0 +1,1215 @@
...
+        table = table.reindex(ix, axis=0)
+        table = table.reindex(ix, axis=1)
+
+    # Ensures that the roes and columns are in the same order.
+    table = table.reindex(table.columns)
+
+    return table
+
+
+class _bunch(object):
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: capitalized `_Bunch` (pep-8)

File: statsmodels/stats/contingency_tables.py [hunk 3]
```diff
@@ -0,0 +1,1215 @@
...
+        df : integer
+            The degrees of freedom of the reference distribution
+        pvalue : float
+            The p-value for the test.
+        """
+        stat = np.asarray(self.chi2_contribs).sum()
+        df = np.prod(np.asarray(self.table.shape) - 1)
+        pvalue = 1 - stats.chi2.cdf(stat, df)
+        b = _bunch()
+        b.stat = stat
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: need to check, I think we (and scipy.stats) spell it out `statistic`    
(consistency across statsmodels is not yet checked and enforced)

File: statsmodels/stats/contingency_tables.py [hunk 4]
```diff
@@ -0,0 +1,1215 @@
...
+        Assess independence between rows and columns for ordinal factors.
+
+        This is the 'linear by linear' association test.
+
+        Parameters
+        ----------
+        row_scores : array-like
+            An array of numeric row scores
+        col_scores : array-like
+            An array of numeric column scores
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: What are "scores"?   I still have problems with different meanings of that word in different contexts.
Are these counts in the context of contingency tables?

File: statsmodels/stats/contingency_tables.py [hunk 5]
```diff
@@ -0,0 +1,1215 @@
...
+
+        Returns a bunch with the following attributes:
+
+        stat : float
+            The test statistic.
+        stat_e0 : float
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: We usually spell out `"null"`, e.g. for likelihood llnull

File: statsmodels/stats/contingency_tables.py [hunk 6]
```diff
@@ -0,0 +1,1215 @@
...
+        sd_stat = np.sqrt(v_stat)
+
+        zscore = (stat - e_stat) / sd_stat
+        pvalue = 2 * stats.norm.cdf(-np.abs(zscore))
+
+        b = _bunch()
+        b.stat = stat
+        b.stat_e0 = e_stat
+        b.stat_sd0 = sd_stat
+        b.zscore = zscore
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: terminology: zscore is specific to normal.

We don't have a fully establishing naming convention for this yet (scipy.stats is also still inconsistent in what the returned `statistic` is, AFAIK)

I'm trying to move towards, `effect`, `effect_size` (standardized effect) and `statistic` for the actual test statistic (i.e. zscore here. But I'm not sure yet about all the different versions of "statistics".

File: statsmodels/stats/contingency_tables.py [hunk 7]
```diff
@@ -0,0 +1,1215 @@
...
+
+
+    @cache_readonly
+    def fittedvalues(self):
+        """
+        Fitted values under independence.
+        """
+        probs = self.independence_probabilities
+        fit = self.table.sum() * probs
+        return fit
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: general question, because I looked at related questions in #2605
independence here is conditional on row and column marginal probabilities.

How does this fit together with contingency tables that 
(1) only have one margin fixed by design, e.g. comparing two independent proportions and 
(2) only have the total sample size fixed by design, e.g. two paired samples as in agreement measures.

(I still only have eclectic knowledge about contingency tables.)

File: statsmodels/stats/contingency_tables.py [hunk 8]
```diff
@@ -0,0 +1,1215 @@
...
+        c = ta[1:, 0:-1]
+        d = ta[1:, 1:]
+        tab = np.log(a) + np.log(d) - np.log(b) - np.log(c)
+        rslt = np.empty(self.table.shape, np.float64)
+        rslt *= np.nan
+        rslt[0:-1, 0:-1] = tab
+
+        if isinstance(self.table_orig, pd.DataFrame):
+            rslt = pd.DataFrame(rslt, index=self.table_orig.index,
+                                columns=self.table_orig.columns)
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: A possible problem of caching the DataFrame is that internal calculations that use this will have either DataFrame or numpy arrays to work with. That makes the code more fragile if there are calculations where the behavior of DataFrames and numpy arrays differ.

File: statsmodels/stats/contingency_tables.py [hunk 9]
```diff
@@ -0,0 +1,1215 @@
...
+        upp_idx = np.triu_indices(k, 1)
+
+        tril = self.table.T[upp_idx]   # lower triangle in column order
+        triu = self.table[upp_idx]     # upper triangle in row order
+
+        stat = ((tril - triu)**2 / (tril + triu + 1e-20)).sum()
+        df = k * (k-1) / 2.
+        pval = stats.chi2.sf(stat, df)
+
+        return stat, pval, df
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: return `_Bunch` for consistency?

File: statsmodels/stats/contingency_tables.py [hunk 10]
```diff
@@ -0,0 +1,1215 @@
...
+
+
+    @cache_readonly
+    def oddsratio_pvalue(self):
+        """
+        P-value for the null hypothesis that the odds ratio equals 1.
+        """
+        zscore = self.log_oddsratio / self.log_oddsratio_se
+        pvalue = 2 * stats.norm.cdf(-np.abs(zscore))
+        return pvalue
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: based on #2605 and the proposed enhancements, I think we will need a `method` argument here so it would need to be a method and not a property.
Assuming we want to add different tests with choices similar to `proportion_confint`.

A `method` argument could be added at some later time, but switching from cached property to method would break backwards compatibility.

(Also: the more general version would allow the null hypothesis that odds ratio is equal to some value different from 1, for inferiority, superiority and similar tests.)

File: statsmodels/stats/contingency_tables.py [hunk 11]
```diff
@@ -0,0 +1,1215 @@
...
+        """
+        f = -stats.norm.ppf(alpha / 2)
+        lor = self.log_oddsratio
+        se = self.log_oddsratio_se
+        lcb = lor - f * se
+        ucb = lor + f * se
+        return lcb, ucb
+
+
+    def oddsratio_confint(self, alpha):
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: we usually define a default `alpha=0.05`

File: statsmodels/stats/contingency_tables.py [hunk 12]
```diff
@@ -0,0 +1,1215 @@
...
+
+
+    @cache_readonly
+    def riskratio_pvalue(self):
+        """
+        p-value for the null hypothesis that the risk ratio equals 1.
+        """
+        zscore = self.log_riskratio / self.log_riskratio_se
+        pvalue = 2 * stats.norm.cdf(-np.abs(zscore))
+        return pvalue
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: same as for odds ratio, a method would be better for future enhancements

File: docs/source/contingency_tables.rst [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+hypotheses that respect the ordering.  The test statistic for the
+linear by linear association test is
+
+.. math::
+
+\sum_k r_i c_j T_{ij}
+
+where :math:`r_i` and :math:`c_j` are row and column scores.  Often
+these scores are set to the sequences 0, 1, ....  This gives the
+'Cochran-Armitage trend test'.
```
Comment (josef-pkt) [inline docs/source/contingency_tables.rst]: scores sounds a bit like weights in cohen's kappa http://statsmodels.sourceforge.net/devel/generated/statsmodels.stats.inter_rater.cohens_kappa.html#statsmodels.stats.inter_rater.cohens_kappa
but I never looked enough at Cochran-Armitage trend test or similar

**Comment by josef-pkt** [comment]
Looks very good to me, brings the things nicely together (and is a bit similar to the weightstats classes)

3 types of comments

- naming convention, API: we still don't have a strict convention for what tests should return
- future proofing: make it easier to add enhancements without breaking backwards compatibility
- conceptual: To me contingency tables look like just a way of arranging the data. I feel more comfortable with looking at the underlying sampling properties, i.e.
  - 1 sample proportions
  - 2 independent samples: table with one margin fixed
  - k independent samples: e.g. stratified 2x2 tables
  - 2 paired sample: Square table with only total nobs fixed.

related is what the tests are conditioning on. (for #2605 I was looking at unconditional exact tests, that only take those numbers fixed that are fixed by the sampling design)

For the current methods and results it would maybe be enough to specify what the conditioning is on. and for stats usage to specify which table and method applies for which sampling case.

File: statsmodels/stats/contingency_tables.py [hunk 14]
```diff
@@ -0,0 +1,1215 @@
...
+        stat /= denom
+
+        # df is always 1
+        pvalue = 1 - stats.chi2.cdf(stat, 1)
+
+        return stat, pvalue
+
+
+    @cache_readonly
+    def common_odds(self):
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: I would prefer postfix `pooled`, e.g. oddsratio_pooled (standard terminology for pooled sample)

**Comment by kshedden** [comment]
@josef-pkt  Thanks, I think I have updated the code to address most of the points you raised.

A few comments:

* We only cache the dataframe so that we can format results analogously to it, we never use it internally for computations (i.e. all calculations are done on the `table` attribute, which is an `ndarray`, not the `table_orig` attribute, which may be a dataframe)

* I think the sampling model for everything here is iid (or for stratified tables, iid within tables, independent between tables).  I haven't implemented anything like Fishers exact test based on fixed margins.  You can think of McNemars as applying to paired data, but you can also think of it as applying to iid data with two attributes per subject.

**Comment by kshedden** [comment]
Seems to be fixed now, the remaining travis errors are in places not touched by this code.

**Comment by josef-pkt** [comment]
@kshedden  This branch looks far behind master, rebase please.

The test results with the newest versions of dependencies look really awful. Hopefully some go away after a rebase. AFAICS, both numpy and pandas are new versions.

I will go over this PR again in the next few days so we can merge this soon.

File: statsmodels/stats/contingency_tables.py [hunk 15]
```diff
@@ -0,0 +1,1312 @@
...
+  * Table : implements methods that can be applied to any two-way
+  contingency table.
+
+  * SquareTable : implements methods that can be applied to a square
+  two-way contingency table.
+
+  * Table2x2 : implements methods that can be applied to a 2x2
+  contingency table.
+
+  * StratifiedTables : implements methods that can be applied to a
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: should `StratifiedTables` be in singular form `StratifiedTable` ?
It's one stratified table?

File: statsmodels/stats/contingency_tables.py [hunk 16]
```diff
@@ -0,0 +1,1312 @@
...
+            statistic : float
+                The chi^2 test statistic.
+            df : integer
+                The degrees of freedom of the reference distribution
+            pvalue : float
+                The p-value for the test.
+    marginal_probabilities : tuple of two arrays
+        The estimated row and column marginal distributions.
+    independence_probabilities : array-like
+        Estimated cell probabilities under independence.
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: conditional on both margins ?

File: statsmodels/stats/contingency_tables.py [hunk 17]
```diff
@@ -0,0 +1,1312 @@
...
+    fittedvalues : array-like
+        Fitted values under independence.
+    pearson_resids : array-like
+        The Pearson residuals.
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: under independence

File: statsmodels/stats/contingency_tables.py [hunk 18]
```diff
@@ -0,0 +1,1312 @@
...
+    standardized_resids : array-like
+        Residuals with approximately unit variance.
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: under independence

Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: I'm not sure anymore we have to specify it for each attribute.

File: statsmodels/stats/contingency_tables.py [hunk 19]
```diff
@@ -0,0 +1,1312 @@
...
+        The local odds ratios are calculated from each 2x2 subtable
+        formed from adjacent rows and columns.
+    cumulative_log_oddsratios : array-like
+        The cumulative log odds ratio at a given pair of thresholds is
+        calculated by reducing the table to a 2x2 table based on
+        dichotomizing the rows and columns at the given thresholds.
+        The table of cumulative log odds ratios presents all possible
+        cumulative log odds ratios that can be formed from a given
+        table.
+    cumulative_oddsratios : array-like
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: `array_like` is not well defined for Returns or attributes, `ndarray` or `DataFrame`?

Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: or scalar in some cases?

In results class we don't specify anything, but I guess we should

File: statsmodels/stats/contingency_tables.py [hunk 20]
```diff
@@ -0,0 +1,1312 @@
...
+        if isinstance(data, pd.DataFrame):
+            table = pd.crosstab(data.iloc[:, 0], data.iloc[:, 1])
+        else:
+            table = pd.crosstab(data[:, 0], data[:, 1])
+
+        return cls(table, shift_zeros)
+
+
+    @cache_readonly
+    def nominal_association(self):
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: two possible problems

- nominal_association might mean a measure for how large the association is (eg. and agreement measure or effectsize), maybe prefix `test_` ?
- making this a cached attribute is not future proof, if for example we want to use likelihood ratio (Q) test or general powerdivergence (Cressie-Reed), or exact tests. 

(I don't remember where I mentioned this before)

File: statsmodels/stats/contingency_tables.py (hunk 4 above)
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: `scores` seems to be commonly used for this across packages

File: statsmodels/stats/contingency_tables.py [hunk 21]
```diff
@@ -0,0 +1,1312 @@
...
+        fit = self.table.sum() * probs
+        return fit
+
+
+    @cache_readonly
+    def pearson_resids(self):
+        # docstring for cached attributes in init above
+
+        fit = self.fittedvalues
+        resids = (self.table - fit) / np.sqrt(fit)
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: I'm puzzled: shouldn't there be a p * (1 - p) in the variance calculation?
Binomial variance?

File: statsmodels/stats/contingency_tables.py [hunk 22]
```diff
@@ -0,0 +1,1312 @@
...
+        return resids
+
+
+    @cache_readonly
+    def standardized_resids(self):
+        # docstring for cached attributes in init above
+
+        row, col = self.marginal_probabilities
+        sresids = self.resids / np.sqrt(np.outer(1 - row, 1 - col))
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: Is self.resids defined? I don't find it right now.

File: statsmodels/stats/contingency_tables.py [hunk 23]
```diff
@@ -0,0 +1,1312 @@
...
+    def fittedvalues(self):
+        # docstring for cached attributes in init above
+
+        probs = self.independence_probabilities
+        fit = self.table.sum() * probs
+        return fit
+
+
+    @cache_readonly
+    def pearson_resids(self):
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: naming: for consistency with GLM/GEE,... and post-fix qualifiers
`resid_pearson`

File: statsmodels/stats/contingency_tables.py (hunk 22 above)
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: Also, I thought Pearson and standardized residuals are the same. GLMResults only have Pearson.

File: statsmodels/stats/contingency_tables.py [hunk 24]
```diff
@@ -0,0 +1,1312 @@
...
+        data = [[fmt % sy.statistic, fmt % sy.pvalue, '%d' % sy.df],
+                [fmt % hm.statistic, fmt % hm.pvalue, '%d' % hm.df]]
+        tab = iolib.SimpleTable(data, headers, stubs, data_aligns="r",
+                                 table_dec_above='')
+
+        return tab
+
+
+
+class Table2x2(SquareTable):
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: a `transpose` method would be useful
given the asymmetry in the interpretation (as in the example in the introduction

should it be in `Table` itself and inherited?

File: statsmodels/stats/contingency_tables.py [hunk 25]
```diff
@@ -0,0 +1,1312 @@
...
+        """
+        f = -stats.norm.ppf(alpha / 2)
+        lrr = self.log_riskratio
+        se = self.log_riskratio_se
+        lcb = lrr - f * se
+        ucb = lrr + f * se
+        return lcb, ucb
+
+
+    def riskratio_confint(self, alpha=0.05):
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: it's not clear to me whether we should already add a `method` keyword. 
I expect confint with better small sample properties will be available for the 0.8 release (or soon after)

File: statsmodels/stats/contingency_tables.py [hunk 26]
```diff
@@ -0,0 +1,1312 @@
...
+    Analyses for a collection of 2x2 stratified contingency tables.
+
+    This class implements the 'Cochran-Mantel-Haenszel' and
+    'Breslow-Day' procedures for analyzing collections of 2x2
+    contingency tables.
+
+    Parameters
+    ----------
+    tables : list
+        A list containing 2x2 contingency tables.
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: This could also be already a 2x2xk ndarray

File: statsmodels/stats/contingency_tables.py [hunk 27]
```diff
@@ -0,0 +1,1312 @@
...
+    This results are based on a sampling model in which the units are
+    independent within and between strata.
+    """
+
+    def __init__(self, tables, shift_zeros=False):
+
+        # Create a data cube
+        tables = [np.asarray(x) for x in tables]
+        table = [x[:, :, None] for x in tables]
+        table = np.concatenate(table, axis=2).astype(np.float64)
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: Since `tables` is not used, I think this should do it also
```
if isinstance(tables, list):
    table = np.array(tables).T
```

File: statsmodels/stats/contingency_tables.py [hunk 28]
```diff
@@ -0,0 +1,1312 @@
...
+
+        if shift_zeros:
+            zx = (table == 0).sum(0).sum(0)
+            ix = np.flatnonzero(zx > 0)
+            if len(ix) > 0:
+                table[:, :, ix] += 0.5
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: this looks the same as

```
if shift_zeros:
    table[table == 0] += 0.5
```

but if table is a view on the original table of the user, then we need a copy if we modify it.

File: statsmodels/stats/contingency_tables.py [hunk 29]
```diff
@@ -0,0 +1,1312 @@
...
+        statistic = statistic**2
+        denom = self._apb * self._apc * self._bpd * self._cpd
+        denom /= (self._n**2 * (self._n - 1))
+        denom = np.sum(denom)
+        statistic /= denom
+
+        # df is always 1
+        pvalue = 1 - stats.chi2.cdf(statistic, 1)
+
+        return statistic, pvalue
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: return Bunch? 
there might be other interesting statistics to attach.

File: statsmodels/stats/contingency_tables.py [hunk 30]
```diff
@@ -0,0 +1,1312 @@
...
+    exact : bool
+        If exact is true, then the binomial distribution will be used.
+        If exact is false, then the chisquare distribution will be
+        used, which is the approximation to the distribution of the
+        test statistic for large sample sizes.
+    correction : bool
+        If true, then a continuity correction is used for the chisquare
+        distribution (if exact is false.)
+
+    Returns a bunch with attributes
```
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: needs a 

```
Returns
----------
```

or this will be parsed as part of params, I guess

**Comment by josef-pkt** [comment]
I went through it again, It looks very good overall, and about ready to merge

The only critical issue are the residuals, like `resid_pearson` naming and unit tests (none as far as I can see)

**Comment by josef-pkt** [comment]
I started to look at an example, first impression

- I think we need `__repr__` and/or `__str__` to the Bunch and classes
- it's confusing that `nominal_association` does not and `ordinal_association` wants a call `()`
- The example from the unit tests get_rdataset("Arthritis", "vcd") is not correctly ordered. Either we need to add some ordering support, or have examples that use ordered category levels. 

I started to read some of Friendly's notes/vignettes for vcd, vcd-extra, and similar lecture notes by others. Some of this could go in follow-up PRs, and is not needed for this PR.

File: statsmodels/stats/contingency_tables.py (hunk 27 above)
Comment (kshedden) [inline statsmodels/stats/contingency_tables.py]: I couldn't get this to work.

File: statsmodels/stats/contingency_tables.py (hunk 25 above)
Comment (kshedden) [inline statsmodels/stats/contingency_tables.py]: Sure, any suggestions on the name?  One of these, maybe?

method="asymptotic"
method="clt"

File: statsmodels/stats/contingency_tables.py (hunk 22 above)
Comment (kshedden) [inline statsmodels/stats/contingency_tables.py]: For tables, these seem to be used differently:

https://onlinecourses.science.psu.edu/stat504/node/86

File: statsmodels/stats/contingency_tables.py (hunk 21 above)
Comment (kshedden) [inline statsmodels/stats/contingency_tables.py]: The Pearson residuals use the Poisson model for standardization.  I'm using this definition:

https://onlinecourses.science.psu.edu/stat504/node/86

File: statsmodels/stats/contingency_tables.py (hunk 27 above)
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: transpose is wrong I guess, my original thought was using rollaxis, but dstack seems to work without adding the the extra axis with `None`

```
>>> li = [10**i +np.arange(6).reshape(2,3) for i in range(4)]
>>> 
>>> np.dstack(li).shape
(2, 3, 4)
```

File: statsmodels/stats/contingency_tables.py (hunk 22 above)
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: Either we need a qualifier for resid_pearson or we need to rethink terminology across models. I was planning to use `resid_pearson` as name for standardized residual in all models where it has unit variance (or constant variance if we leave out dispersion).

If it depends on what sampling distributions we assume, Poisson/loglinear, multinomial or binomial, then we need to be more specific when we use which sampling.

I was reading up a bit on the GLM-loglinear examples by Friendly and by others, but still have only a vague overview.

the psu stat504 site looks interesting and useful.

File: statsmodels/stats/contingency_tables.py (hunk 8 above)
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: The dataframe looks a bit strange, in the 2x3 table we have only two values and still the original row and column labels. i.e. the labels don't indicate which oddsratio it is.
cumulative is defined at the lower end of the 2x2 subtable, like cdf I would find upper (in increasing values) corner more intuitive.

Also, shouldn't the comparison in the `nominal` case be "all-pairs" instead of only neighboring pairs.

(I have no prior information, I've never used things like this before.)

File: statsmodels/stats/contingency_tables.py (hunk 25 above)
Comment (josef-pkt) [inline statsmodels/stats/contingency_tables.py]: In proportion_confint I just used `method='normal'`
http://statsmodels.sourceforge.net/devel/generated/statsmodels.stats.proportion.proportion_confint.html#statsmodels.stats.proportion.proportion_confint

**Comment by josef-pkt** [comment]
I'm fine with merging when you are done with this round.

I'll send a message to the mailing list for feedback after merge, and we can review some pieces again later or in the context of #2605 (two sample proportions)

Also, now that I have a bit the connection I'd like to look into the relationship with loglinear/Poisson models, similar for #2605 I wanted to see how much we can replicate and have more general models with GLM Binomial (without any `exact` part).

**Comment by josef-pkt** [comment]
merged, Thanks @kshedden 

we will need to start a changelog for 0.8 to add big enhancements, and others
